tools/bapdoc.ml

```diff
@@ -0,0 +1,232 @@
+security analysis, program verification, and reverse engineering.
+
+The framework consists of a bunch of libraries, plugins and
+frontends. The libraries provide code reusability, plugins facilitate
+extensibility and frontends serve as entry points.
+
+Frontends come with comprehensive manuals,
+that can be accessed by using [--help] command line options, or via
+the [man] command, if the manpath is configured correctly.
+Finally, you can access a man page for a plugin using
+[--<PLUGIN>-help] command line option of a frontend, e.g., [bap --map-terms-help].
+
+|}
+
+let packages =
+  List.concat_map ~f:(fun (_,ps) -> List.map ~f:(fun (p,_,_) -> p) ps) libraries
+
+let mkdir path =
+  if not (Sys.file_exists path) then
+    Unix.mkdir path 0o770
+
+let run cmd =
+  let res = Sys.command cmd in
+  if res <> 0 then
+    failwith ("Command: '" ^ cmd ^ "' failed")
+
+let dedot name =
+  match String.split name ~on:'.' with
+  | name' :: _ when name' = "Bap" -> name
+  | name :: _ -> name
+  | _ -> name
+
+let render_entry (_,entry,desc) =
+  sprintf "- {{!%s}%s} - %s" entry (dedot entry) desc
+
+let render_section (name,entries) =
+  sprintf "{1 %s}\n%s" name
+    (String.concat ~sep:"\n" @@ List.map ~f:render_entry entries)
+
+type info = {
+  man : string -> string;
+  help : string -> string;
+}
+
+let program = {
+  man  = ident;
+  help = sprintf "%s --help=groff"
+}
+
+let plugin = {
+  man = sprintf "bap-plugin-%s";
+  help = sprintf "bap --%s-help=groff";
+}
+
+let man3_redirection lib =
+  mkdir "man3";
+  let redirection = sprintf {|
+    <!doctype html>
+    <html>
+    <head>
+    <meta http-equiv="refresh" content="0; url='../odoc/%s/index.html'" />
+    </head>
+    <body></body>
+    </html> |} lib in
+  let name = sprintf "man3/%s.3.html" lib in
+  Out_channel.write_all name ~data:redirection
+
+let build_manual {man; help} tool =
+  let repair_links x =
+    sprintf {|%s | sed "s/\\\N'45'/-/g" | man2html -r > %s|}
+      (help tool) x in
+  let redirect_index x =
+    sprintf {|sed "s#../index.html#../odoc/index.html#g" -i %s|} x in
+  mkdir "man1";
+  match Sys.command @@ sprintf "%s >/dev/null" (help tool) with
+  | 0 ->
+    let output = sprintf "man1/%s.1.html" (man tool) in
+    run @@ repair_links output;
+    run @@ redirect_index output
+  | _ ->
+    eprintf "Warning: can't render manpage for %s\n" tool;
+    eprintf "Called as %S, got:\n" (help tool);
+    ignore(Sys.command @@ sprintf "%s >&2" (help tool));
+    eprintf "\n%!"
+
+let generate_manual () =
+  let _frontends =
+    List.iter frontends ~f:(fun (p,_) -> build_manual program p) in
+  let _plugins =
+    Plugins.list () |> List.iter ~f:(fun p ->
+        build_manual plugin (Plugin.name p))in
+  List.iter packages ~f:(fun lib -> man3_redirection lib)
+
+let odig_pkgs () =
+  let f = Filename.temp_file "odig" ".pkgs" in
+  run (sprintf "odig pkg > %s" f);
+  let pkgs =
+    In_channel.read_lines f |>
+    List.filter_map ~f:(fun s ->
```

Comment:
refactor it into the `remove_unresolved_pkgs` or `is_resolved_pkg` if you will refactor only the predicate.